packages/good-design/src/core/buttons/ClaimButton.tsx

```diff
@@ -134,6 +134,17 @@ const ClaimButton = ({
       setFirstClaim(first);
       showActionModal();
 
+      // if handleModalOpen is triggered, and isWhitelisted is undefined
+      // it means a wallet used the claim button to connect and the blockchain data has not been fetched yet
+      // so
+      // 1. we should retry after x seconds by running a seperate useEffect and a awaitingWhitelist state variable
+      // 2. isWhitelisted should be returned as Promise<boolean> to be awaited on
```

Comment:
@L03TJ3 
you could use them both - some bool state var (e.g. isWhitelisted) and some async function to re-request status if isWhitelisted is still false

Here is an example from the wallet's claim button:


```javascript
const { dailyUBI: entitlement, isCitizen } = useWalletData()

// ...
const onClaim = useCallback(async () => {
    let curEntitlement
    let isWhitelisted = isCitizen

    setLoading(true)

    // 1. Pre-checks before claim. Wrapping / retrying / handling errors separately
    try {
      await _retry(async () => {
        // Call wallet method to refresh the connection, silent fail
        await goodWallet.getBlockNumber().catch(e => log.warn('getBlockNumber failed', e.message, e))

        // recheck citizen status, just in case we are out of sync with blockchain
        if (!isCitizen) {
          isWhitelisted = await goodWallet.isCitizen()
        }

        if (isWhitelisted) {
          // when we come back from FR entitlement might not be set yet
          curEntitlement = dailyUbi || (await goodWallet.checkEntitlement().then(parseInt))
        }
      })

      // ...
```

Reply:
After a second glance I went with a (IMO) simple extra useEffect.